File: sdk/ios/Sources/OpenStorySDK/Internal/Context/UserContextStore.swift
import Foundation

internal final class UserContextStore: @unchecked Sendable {
    private let lock = NSLock()
    private var userSegments: [String] = []

    func update(_ segments: some Collection<String>) {
        let normalized = UserSegmentsNormalizer.normalize(segments)
        lock.lock()
        userSegments = normalized
        lock.unlock()
    }

    func snapshot() -> [String] {
        lock.lock()
        defer { lock.unlock() }
        return userSegments
    }
}

File: sdk/flutter/android/src/main/kotlin/com/openstory/sdk/internal/context/UserContextStore.kt
package com.openstory.sdk.internal.context

import java.util.concurrent.atomic.AtomicReference

internal class UserContextStore {
    private val segmentsRef = AtomicReference<List<String>>(emptyList())

    fun update(userSegments: Collection<String>) {
        segmentsRef.set(UserSegmentsNormalizer.normalize(userSegments))
    }

    fun snapshot(): List<String> = segmentsRef.get()
}
